[2.6] Add notifiers to the notifications guide (#2181)
astronomer/docs: learn/error-notifications-in-airflow.md

Any Python callable can be provided to the `*_callback` parameters. Airflow 2.6 added the functionality of using [notifiers](#) and the option to provide several callback items to the same callback parameter in a list.

:::info

The number of retries a task will be given can be set using the `retries` parameter. The time spent between retries is defined by the `retry_delay` (default: `timedelta(seconds=300)`). To progressively increases the wait time between retries until `max_retry_delay` (default: `None`) is reached, set `retry_exponential_backoff` to `True`.

:::

## Notification levels

To define a notification at the DAG level, you can set the `*_callback` parameters in your DAG instantiation.

```python
def my_callback_function(context):
    pass

@dag(
    start_date=datetime(2023,4,25),
    schedule="@daily",
    catchup=False,
    on_success_callback=my_callback_function,
    on_failure_callback=my_callback_function,
    sla_miss_callback=my_callback_function

> )

Reviewer: We don't have to actually write all of these functions, we could just use a `...` in DAG code to imply that they exist

<Tabs
    defaultValue="taskflow"
    groupId="notification-levels"
    values={[
        {label: 'TaskFlow API', value: 'taskflow'},
        {label: 'Traditional syntax', value: 'traditional'},
    ]}>

<TabItem value="taskflow">

```python
def my_callback_function(context):
    pass

@task(
    on_execute_callback=my_callback_function,
    on_retry_callback=my_callback_function,
    on_success_callback=my_callback_function,
    on_failure_callback=my_callback_function,
)
def t1():
    return "hello"
```

</TabItem>
<TabItem value="traditional">

```python
def my_callback_function(context):
    pass

def say_hello():
    return "hello"

t1 = PythonOperator(
    task_id="t1",
    python_callable=say_hello,
    on_execute_callback=my_callback_function,
    on_retry_callback=my_callback_function,
    on_success_callback=my_callback_function,
    on_failure_callback=my_callback_function,
)
```

</TabItem>
</Tabs>

## Email notifications

> Aside from the general purpose `*_callback` parameters, all Airflow operators are able to use the two BaseOperator parameters `email_on_failure` and `email_on_retry` if the parameter `email` has been set and the `SMTP` have been configured for the Airflow environment. 

Reviewer: ```suggestion
If you have an SMTP connection configured in Airflow, you can use the `email, `email_on_failure`, and `email_on_retry` task parameters to send notification emails from Airflow.
```

<Tabs
    defaultValue="taskflow"
    groupId="email-notifications"
    values={[
        {label: 'TaskFlow API', value: 'taskflow'},
        {label: 'Traditional syntax', value: 'traditional'},
    ]}>

<TabItem value="taskflow">

```python
@task(
    email=["noreply@astronomer.io", "noreply2@astronomer.io"],
    email_on_failure=True,
    email_on_retry=True
)
def t1():
    return "hello"
```

</TabItem>
<TabItem value="traditional">

```python
def say_hello():
    return "hello"

t1 = PythonOperator(
    task_id="t1",
    python_callable=say_hello,
    email=["noreply@astronomer.io", "noreply2@astronomer.io"],
    email_on_failure=True,
    email_on_retry=True
)
```

</TabItem>
</Tabs>

> The email notification parameters can be given to all tasks in an Airflow DAG by defining them in the `default_args` parameter.

Reviewer: ```suggestion
You can also configure email notifications for all tasks in a DAG by defining the configurations in the `default_args` parameter.
```

```python
default_args = {
    "email": ["noreply@astronomer.io"],
    "email_on_failure": True,
    "email_on_retry": True,
}

@dag(
    start_date=datetime(2023, 4, 25),
    schedule="@daily",
    catchup=None,
    default_args=default_args
)
```

> To allow Airflow to send emails, you have to provide values to the SMTP section of your `airflow.cfg` similar to this example:

Reviewer: Can you also just configure a connection from the Airflow UI?

Reviewer 2: I would not mention this. I think it's adequate to link to the airflow docs as you already do. And in my experience it's much more common to configure smtp with env vars

> ## Airflow notification basics

Reviewer: ```suggestion
## Airflow notification types
```

Great use case breakdown here

> When using Airflow, you have different options to configure notifications geared towards different use cases:

Reviewer: ```suggestion
Airflow has a few options for notifying you on the status of your DAGs:
```

> - **Email notifications**: Every Airflow operator derived from the BaseOperator has parameters dedicated to configure email alerts in case of a task failure or retry. Email alerts are often set on tasks in production pipelines where task failures or retries need immediate attention by a data professional. 

Reviewer: ```suggestion
- **Email notifications**: Most Airflow operators have parameters for setting email alerts in case of a task failure or retry. Use email alerts in production pipelines where task failures or retries need immediate attention by a data professional. 
```

> - **Airflow callbacks**: Callback parameters (`*_callback`) exist both at the task and at the DAG level. You can pass any callable or Airflow notifier to these parameters to be executed in different events like success or failure. Airflow callbacks offer a lot of flexibility to execute any code based on the state of a task or DAG. They are often used to define actions specific to a task or DAG failure or success. 

Reviewer: ```suggestion
- **Airflow callbacks**: Callback parameters (`*_callback`) exist both at the task and at the DAG level. You can pass any callable or Airflow notifier to these parameters, and Airflow will run them in the case of specific events, such as a task failure. Airflow callbacks offer a lot of flexibility to execute any code based on the state of a task or DAG. They are often used to define actions for specific instances of task failures or successes.
```

> - **Airflow notifiers**: Notifiers, which were added in Airflow 2.6, offer a way to modularize callback functions into custom classes to easily be reused and standardized. Provider packages can ship pre-built notifiers like the SlackNotifier. Notifiers can be provided to all callback parameters to define which task or DAG state should cause them to be executed. A common use case for notifiers is standardizing the action to be taken on task failures across several Airflow instances.

Reviewer 2: Link to SlackNotifier docs on that word? (if there are any?....seems like this should go on the registry)

> - **Airflow notifiers**: Notifiers, which were added in Airflow 2.6, offer a way to modularize callback functions into custom classes to easily be reused and standardized. Provider packages can ship pre-built notifiers like the [SlackNotifier](https://registry.astronomer.io/providers/apache-airflow-providers-slack/versions/7.2.0/modules/SlackNotifier). Notifiers can be provided to all callback parameters to define which task or DAG state should cause them to be executed. A common use case for notifiers is standardizing the action to be taken on task failures across several Airflow instances.

Reviewer: ```suggestion
- **Airflow notifiers**: Notifiers are custom classes for callback functions that can be easily reused and standardized. Provider packages can ship pre-built notifiers like the [SlackNotifier](https://registry.astronomer.io/providers/apache-airflow-providers-slack/versions/7.2.0/modules/SlackNotifier). Notifiers can be provided to callback parameters to define which task or DAG state should cause them to be executed. A common use case for notifiers is standardizing actions for task failures across several Airflow instances.
```

> - **Airflow service-level agreements (SLAs)**: SLAs in Airflow define the expected time it takes from the DAG execution until a specific task has been completed. If an SLA is missed the callable or notifier provided to the `sla_miss_callback` parameter is executed and if the SMTP connection is configured an email will be sent as well. Since an SLA miss does not stop a task from running, this type of notification is used when intervention by a data professional is needed if a specific task is taking longer than expected.

Reviewer 2: ```suggestion
- **Airflow service-level agreements (SLAs)**: SLAs in Airflow define the expected time it takes from the DAG execution until a specific task has been completed. If an SLA is missed, the callable or notifier provided to the `sla_miss_callback` parameter is executed. If an SMTP connection is configured in that Airflow environment, an email will be sent as well. Since an SLA miss does not stop a task from running, this type of notification is used when intervention is needed if a specific task is taking longer than expected.
```

> - **Airflow service-level agreements (SLAs)**: SLAs in Airflow define the expected time it takes from the DAG execution until a specific task has been completed. If an SLA is missed, the callable or notifier provided to the `sla_miss_callback` parameter is executed. If an SMTP connection is configured in that Airflow environment, an email will be sent as well. Since an SLA miss does not stop a task from running, this type of notification is used when intervention is needed if a specific task is taking longer than expected.

Reviewer: ```suggestion
- **Airflow service-level agreements (SLAs)**: SLAs define the expected time it takes for a specific task to complete. If an SLA is missed, the callable or notifier provided to the `sla_miss_callback` parameter is executed. If you configure an SMTP connection, an email will be sent as well. Since an SLA miss does not stop a task from running, this type of notification is used when intervention is needed if a specific task is taking longer than expected.
```

> Most notifications can be set at the level of both a DAG and a task, like the `on_failure_callback`, while others can only be set at the task or DAG level. Setting a parameter within a DAGs `default_args` dictionary will apply it to all tasks in the DAG, you can see examples in the [set DAG and task level callbacks](#set-dag-and-task-level-callbacks) section.

Reviewer 2: ```suggestion
Most notifications can be set at the level of both a DAG and a task. Setting a parameter within a DAG's `default_args` dictionary will apply it to all tasks in the DAG. You can see examples of this in the [set DAG and task level callbacks](#set-dag-and-task-level-callbacks) section.
```

You can also set these values using environment variables. In this case, all parameters are preceded by `AIRFLOW__SMTP__`. For example, `smtp_host` can be specified by setting the `AIRFLOW__SMTP__SMTP_HOST` variable. For more on Airflow email configuration, see [Email Configuration](https://airflow.apache.org/docs/apache-airflow/stable/howto/email-config.html). 

> If you are using Astro, use environment variables to set up SMTP because the `airflow.cfg` cannot be directly edited.

Reviewer 2: Is there a doc we can link to for this? (Astro env vars?)

If you want to send emails out on a more customizable basis, you can also use Airflow's callback functions to run custom functions that send email notifications. For example, if you want to send emails for successful task runs, you can provide an email function to the `on_success_callback` parameter:

```python
from airflow.utils.email import send_email

def success_email_function(context):
    dag_run = context.get("dag_run")

    msg = "DAG ran successfully"
    subject = f"DAG {dag_run} has completed"
    send_email(to=your_emails, subject=subject, html_content=msg)

@dag(
    start_date=datetime(2023, 4, 26),
    schedule="@daily",
    catchup=False,
    on_success_callback=success_email_function
)
```

## Airflow callbacks 

In Airflow you can define actions to be taken due to different DAG or task states using `*_callback` parameters:

- `on_success_callback`: Invoked when a task or DAG succeeds.
- `on_failure_callback`: Invoked when a task or DAG fails.
- `on_execute_callback`: Invoked right before a task begins executing. This callback only exists at the task level.
- `on_retry_callback`: Invoked when a task is retried. This callback only exists at the task level.

> - `sla_miss_callback`: Invoked when a task or DAG misses its defined [Service Level Agreement (SLA)](#airflow-service-level-agreements). This callback can only be defined at the DAG level to be used for each task in a DAG which has an SLA defined.

Reviewer 2: ```suggestion
- `sla_miss_callback`: Invoked when a task or DAG misses its defined [Service Level Agreement (SLA)](#airflow-service-level-agreements). This callback is defined at the DAG level for DAGs with defined SLAs and will be applied to every task.
```

> Any Python callable can be provided to the `*_callback` parameters. Airflow 2.6 added the functionality of using [notifiers](#notifiers) and the option to provide several callback items to the same callback parameter in a list.

Reviewer 2: ```suggestion
You can provide any Python callable to the `*_callback` parameters. As of Airflow 2.6, you can also use [notifiers](#notifiers) for your callbacks, and you can provide several callback items to the same callback parameter in a list.
```
Not sure this is really better wording, but trying to make it more consistent.

> ### Set DAG and task level callbacks

Reviewer: ```suggestion
### Set DAG and task-level callbacks
```

```python
from airflow.utils.email import send_email

def success_email_function(context):
    dag_run = context.get("dag_run")

    msg = "DAG ran successfully"
    subject = f"DAG {dag_run} has completed"
    send_email(to=your_emails, subject=subject, html_content=msg)

@dag(
    start_date=datetime(2023, 4, 26),
    schedule="@daily",
    catchup=False,
    on_success_callback=success_email_function
)
```

## Airflow callbacks 

In Airflow you can define actions to be taken after successful and failed runs using the following parameters:

- `on_success_callback`: Invoked when a task or DAG succeeds.
- `on_failure_callback`: Invoked when a task or DAG fails.

Additionally there are two task parameters that take a callable which is executed when a task is in a certain state:

- `on_execute_callback`: Invoked right before a task begins executing.
- `on_retry_callback`: Invoked when a task is retried. 

> The `sla_miss_callback` which is invoked when a task or DAG misses its defined [Service Level Agreement (SLA)](#airflow-service-level-agreements) can only be defined at the DAG level to be used for each task in a DAG which has an SLA defined.

Reviewer 2: ```suggestion
The `sla_miss_callback` which is invoked when a task or DAG misses its defined [Service Level Agreement (SLA)](#airflow-service-level-agreements) can only be defined at the DAG level to be used for each task in a DAG which has an SLA defined.
```
I would lean towards putting all 5 callbacks in one bulleted list along with their descriptions. But open to opposing feedback.

Any Python callable can be provided to the `*_callback` parameters. Airflow 2.6 added the functionality of using [notifiers](#notifiers) and the option to provide several callback items to the same callback parameter in a list.

### Set DAG and task-level callbacks

> To define a notification at the DAG level, you can set the `*_callback` parameters in your DAG instantiation.

Reviewer 2: Let's add a sentence here just to be super clear that "DAG level callbacks will send a notification based on the state of the entire DAG run" or something like that. to differentiate from a default arg

```python
def my_success_callback_function(context):
    pass

def my_failure_callback_function(context):
    pass

def my_sla_callback_function(context):
    pass

@dag(
    start_date=datetime(2023,4,25),
    schedule="@daily",
    catchup=False,
    on_success_callback=my_success_callback_function,
    on_failure_callback=my_failure_callback_function,
    sla_miss_callback=my_sla_callback_function
)
```

> If you want each task in a DAG to invoke a task-level call back you can standarize task-level callbacks across your entire DAG using the DAG parameter `default_args`. Items listed in the dictionary provided to the `default_args` parameter will be set for each task in the DAG.

Reviewer 2: ```suggestion
To apply a task-level callback to each task in your DAG, you can pass the callback function to the `default_args` parameter. Items listed in the dictionary provided to the `default_args` parameter will be set for each task in the DAG.
```

You can customize this content by setting the `subject_template` and/or `html_content_template` variables in your `airflow.cfg` with the path to your jinja template files for subject and content respectively.

> Custom callback functions can also be used to send email notifications. For example, if you want to send emails for successful task runs, you can provide an email function to your `on_success_callback`:

Reviewer: ```suggestion
If you want to send emails out on a more customizable basis, you can also use Airflow's callback functions to run custom functions that send email notifications. For example, if you want to send emails for successful task runs, you can provide an email function to the `on_success_callback` parameter:
```

```python
from airflow.utils.email import send_email

def success_email_function(context):
    dag_run = context.get("dag_run")

    msg = "DAG ran successfully"
    subject = f"DAG {dag_run} has completed"
    send_email(to=your_emails, subject=subject, html_content=msg)

```

## Notifiers

> Airflow 2.6 added the concept of [notifiers](https://airflow.apache.org/docs/apache-airflow/stable/howto/notifications.html), which are pre-built or custom classes and can be used to standardize and modularize callback functions.

Reviewer: Maybe one more sentence here that makes explicit the difference between notifiers and notifications? It's that you can pass values into the functions?

Reviewer 2: Agree we should add osme additional context here. I would add an extra sentence making it clear that notifiers are defined globally in your Airflow environment and can be used across any of your DAGs, and that their extensibility means that the community can add them to providers so everybody can use them (this is the biggest benefit I see of them - essentially they provide a framework for pre-built callback functions that anybody can contribute to Airflow, so you might not have to define your own. They're also a bit easier to use across DAGs than an old callback function would be).

And to Jake's question - maybe in the first sentence say "standardize and modularize the functions you use to send notifications. Notifiers get passed to the relevant on_*_callback function of your DAG depending on what event you are notifying on." would be clearer? 

In general, a notifier is a function that sends a notification for you. Also, a callback function doesn't necessarily need to be a notification.

An Airflow notifier can be created by inheriting from the `BaseNotifier` class and defining the action which should be taken in case the notifier is used in the `.notify()` method.

```python
class MyNotifier(BaseNotifier):
    """
    Basic notifier, prints the task_id, state and a message.
    """

    template_fields = ("message",)

    def __init__(self, message):
        self.message = message

    def notify(self, context):
        t_id = context["ti"].task_id
        t_state = context["ti"].state
        print(
            f"Hi from MyNotifier! {t_id} finished as: {t_state} and says {self.message}"
        )
```

> To use the custom notifier in a DAG provide its instantiation to any callback parameter, for example:

Reviewer: ```suggestion
To use the custom notifier in a DAG, provide its instantiation to any callback parameter. For example:
```

<Tabs
    defaultValue="taskflow"
    groupId="notifiers"
    values={[
        {label: 'TaskFlow API', value: 'taskflow'},
        {label: 'Traditional syntax', value: 'traditional'},
    ]}>

<TabItem value="taskflow">

```python
@task(
    on_failure_callback=MyNotifier(message="Hello failed!"),
)
def t1():
    return "hello"
```

</TabItem>
<TabItem value="traditional">

```python
def say_hello():
    return "hello"

t1 = PythonOperator(
    task_id="t1",
    python_callable=say_hello,
    on_failure_callback=MyNotifier(message="Hello failed!"),
)
```

</TabItem>
</Tabs>

> Notifiers are useful to abstract and standardize complex logic used to send a notification from Airflow. 

Reviewer: Isn't this also true of the standard callback functions?

Reviewer 2: I think we can remove this sentence in favor of the additions to the intro to this section mentioned above.

And this isn't _that_ true of standard callback functions. Those are more "define whatever you want" but there's no structure at all. Notifiers provide more of a framework so it's a first-class feature. And while you still _could_ define whatever you want, the biggest benefit will be the community taking advantage of the framework and contributing system-specific notifiers.

### Pre-built notifier: Slack

An example of a community provided pre-built notifier is the [SlackNotifier](https://airflow.apache.org/docs/apache-airflow-providers-slack/latest/_api/airflow/providers/slack/notifications/slack_notifier/index.html). 

It can be imported from the Slack provider package and used with any `*_callback` function:

<CodeBlock language="python">{slack_notifier_example_dag}</CodeBlock>

> The DAG above has one task sending a notification to Slack. It necessitates configuring an [Airflow connection](connections.md) to Slack with the connection ID `slack_conn`.

Reviewer: ```suggestion
The DAG above has one task sending a notification to Slack. It uses a Slack [Airflow connection](connections.md) with the connection ID `slack_conn`.
```

<CodeBlock language="python">{sla_task_level_example_taskflow}</CodeBlock>

</TabItem>
<TabItem value="traditional">

> <CodeBlock language="python">{sla_task_level_example_traditional}</CodeBlock>

Reviewer: `![SLA Email](/img/guides/sla_email.png)` looks a bit off in the guide because it blends in with the background of light mode, can we replace it with this version that has a single-pixel black border?
![sla_email](https://user-images.githubusercontent.com/74574233/234536954-ecd9fe43-940b-4e95-b96c-fc66eb3650c1.png)

If you are running Airflow with Astronomer Software or Astro, there are a number of options available for managing your Airflow notifications. All of the previous methods for sending task notifications from Airflow can be implemented on Astronomer. See the [Astronomer Software](https://docs.astronomer.io/software/airflow-alerts) and [Astro](https://docs.astronomer.io/astro/airflow-alerts) documentation to learn how to leverage notifications on the platform, including how to set up SMTP to enable email notifications.

Astronomer also provides deployment and platform-level alerting to notify you if any aspect of your Airflow or Astronomer infrastructure is unhealthy. For more on that, including how to customize notifications for Software users, see [Alerting in Astronomer Software](https://docs.astronomer.io/software/platform-alerts).

## Legacy: Slack notifications pre-2.6

:::info

In Airflow 2.6+ you can use the [SlackNotifier](#pre-built-notifier-slack) to send notifications to Slack from within Airflow in a modularized and standardized way.

:::

> If you are on an Airflow version older than 2.6 and want to implement Slack notification you can follow this example.

Reviewer: ```suggestion
If you are on an Airflow version older than 2.6 and want to implement Slack notifications, follow this example.
```

It uses the [Slack provider](https://registry.astronomer.io/providers/slack) `SlackWebhookOperator` with a Slack Webhook to send messages. This is the method Slack recommends to post messages from apps.

1. From your Slack workspace, create a Slack app and an incoming Webhook. See [Sending messages using Incoming Webhooks](https://api.slack.com/messaging/webhooks). 
2. Copy the Slack Webhook URL. You'll use it in your Python function.
3. Create an Airflow connection to provide your Slack Webhook to Airflow. Choose an HTTP connection type. Enter [`https://hooks.slack.com/services/`](https://hooks.slack.com/services/) as the Host, and enter the remainder of your Webhook URL from the last step as the Password (formatted as `T00000000/B00000000/XXXXXXXXXXXXXXXXXXXXXXXX`). 

> ![Slack Connection](/img/guides/slack_webhook_connection.png)
> 
> In Airflow 2.0 or later, you'll need to install the `apache-airflow-providers-slack` provider package to use the `SlackWebhookOperator`.

Reviewer: ```suggestion
    ![Slack Connection](/img/guides/slack_webhook_connection.png)

    In Airflow 2.0 or later, you'll need to install the `apache-airflow-providers-slack` provider package to use the `SlackWebhookOperator`.
```

4. Create a Python function to use as your `on_failure_callback` method. Within the function, define the information you want to send and invoke the `SlackWebhookOperator` to send the message similar to this example:

> ```python

Reviewer: Indent this code block by 4 spaces
